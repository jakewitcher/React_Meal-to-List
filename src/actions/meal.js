import uuid from 'uuid';

export const addMeal = (
    {
        name = '',
        itemList = [],
    } = {}
) => ({
    type: 'ADD_MEAL',
    meal: {
        id: uuid(),
        name,
        itemList,
    }
});

export const deleteMeal = (id = '') => ({
    type: 'DELETE_MEAL',
    id,
});

export const editMeal = (
    {
        name = '',
        id = '',
        itemList = [],
    } = {}
) => ({
    type: 'EDIT_MEAL',
    meal: {
        id,
        name,
        itemList,
    }
});