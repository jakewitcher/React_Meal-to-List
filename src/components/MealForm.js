import React from "react";

const MealForm = props => {
  const {
    mealNameChange,
    itemNameChange,
    amountChange,
    unitChange,
    addItem,
    resetMeal,
    mealName,
    itemName,
    amount,
    unit,
    itemList,
    updateMeal,
    mealId
  } = props;

  const pluralUnitValue = (amount, unitType) => {
    if (amount < 2) {
      return `${unitType}`;
    }
    return `${unitType}s`;
  };

  const handleMealChange = (name, id, itemList) => {
    updateMeal({
      name,
      itemList,
      id
    });
    resetMeal();
  };

  return (
    <div className="form">
      <div className="input-box input-box--item-name">
        <p>Item</p>
        <input
          className="input"
          type="text"
          value={itemName}
          onChange={itemNameChange}
        />
      </div>

      <div className="input-box input-box--amount">
        <p>Amount</p>
        <input
          className="input"
          type="number"
          value={amount}
          onChange={amountChange}
        />
      </div>

      <div className="input-box input-box--unit">
        <p>Unit</p>
        <select
          className="input input--dropdown"
          name="unit"
          value={unit}
          onChange={unitChange}
        >
          <option value={""}>{"none"}</option>
          <option value={pluralUnitValue(amount, "pound")}>
            {pluralUnitValue(amount, "pound")}
          </option>
          <option value={pluralUnitValue(amount, "ounce")}>
            {pluralUnitValue(amount, "ounce")}
          </option>
          <option value={pluralUnitValue(amount, "gallon")}>
            {pluralUnitValue(amount, "gallon")}
          </option>
          <option value={pluralUnitValue(amount, "quart")}>
            {pluralUnitValue(amount, "quart")}
          </option>
          <option value={pluralUnitValue(amount, "pint")}>
            {pluralUnitValue(amount, "pint")}
          </option>
          <option value={pluralUnitValue(amount, "item")}>
            {pluralUnitValue(amount, "item")}
          </option>
          <option value={pluralUnitValue(amount, "bag")}>
            {pluralUnitValue(amount, "bag")}
          </option>
          <option value={pluralUnitValue(amount, "container")}>
            {pluralUnitValue(amount, "container")}
          </option>
        </select>
      </div>
      <div className="button-box__add">
        <button className="button button__add" onClick={addItem}>
          Add Item
        </button>
      </div>
      <div>
        <div className="input-box__meal-name">
          <p>Meal Name</p>
          <input
            className="input__meal-name"
            type="text"
            value={mealName}
            onChange={mealNameChange}
          />
        </div>
        <div className="button-box__save">
          <button
            className="button button__save"
            onClick={() => handleMealChange(mealName, mealId, itemList)}
          >
            Save Meal
          </button>
        </div>
      </div>
    </div>
  );
};

export default MealForm;
