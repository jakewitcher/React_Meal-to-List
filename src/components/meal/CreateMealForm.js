import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        dispatch
    } = props.mealProps;
    const { addMeal } = props;

    return (
        <div className="form">
            <h2 className="form__header">Create a New Meal</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value="pound(s)">pound(s)</option>
                            <option value="ounce(s)">ounce(s)</option>
                            <option value="gallon(s)">gallons</option>
                            <option value="quart(s)">quart(s)</option>
                            <option value="pint(s)">pint(s)</option>
                            <option value="item(s)">item(s)</option>
                            <option value="bag(s)">bag(s)</option>
                            <option value="box(es)">box(es)</option>
                            <option value="container(s)">container(s)</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            dispatch(addMeal({
                                name,
                                itemList
                            }));
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}



export default MealForm;