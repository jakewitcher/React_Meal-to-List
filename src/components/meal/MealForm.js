import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        updateMeal,
        mealId,
        title,
    } = props;

    const pluralUnitValue = (amount, unitType) => {
        if (amount < 2) {
            return `${unitType}`
        }
        return `${unitType}s`
    };

    return (
        <div className="form">
            <h2 className="form__header">{title}</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value={pluralUnitValue(amount, 'pound')}>{pluralUnitValue(amount, 'pound')}</option>
                            <option value={pluralUnitValue(amount, 'ounce')}>{pluralUnitValue(amount, 'ounce')}</option>
                            <option value={pluralUnitValue(amount, 'gallon')}>{pluralUnitValue(amount, 'gallon')}</option>
                            <option value={pluralUnitValue(amount, 'quart')}>{pluralUnitValue(amount, 'quart')}</option>
                            <option value={pluralUnitValue(amount, 'pint')}>{pluralUnitValue(amount, 'pint')}</option>
                            <option value={pluralUnitValue(amount, 'item')}>{pluralUnitValue(amount, 'item')}</option>
                            <option value={pluralUnitValue(amount, 'bag')}>{pluralUnitValue(amount, 'bag')}</option>
                            <option value={pluralUnitValue(amount, 'container')}>{pluralUnitValue(amount, 'container')}</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            const id = mealId;
                            updateMeal({
                                name,
                                itemList,
                                id,
                            });
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}

export default MealForm;