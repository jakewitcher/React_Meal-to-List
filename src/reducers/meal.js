import { defaultMealsList } from './seed';

// const mealDefaultState = {
//     mealList: [],
// };

export default (state = {mealList: defaultMealsList}, action) => {
    switch (action.type) {
        case 'ADD_MEAL':
            return {
                ...state,
                mealList: [...state.mealList, action.meal]
            };
        default:
            return state;
    }
}