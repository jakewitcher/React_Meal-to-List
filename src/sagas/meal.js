import { put, takeEvery, select } from 'redux-saga/effects';
import database from '../firebase/firebase';

const getUserId = (state) => state.auth.uid;

const formatItemList = (list) => {
    return list.reduce((list, item) => {
        list[item.itemName] = {
            amount: item.amount,
            unit: item.unit
        };
        return list;
    }, {});
};

function* addMeal({ meal = {} } = {}) {
    const { itemList, name } = meal;
    const items = formatItemList(itemList);

    let id;
    const uid = yield select(getUserId);
    yield database.ref(`users/${uid}/meals`).push({ name, items }).then((ref) => {
        id = ref.key;
    });

    yield put({
        type: 'ADD_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
}

function* editMeal({ meal = {} }) {
    const { itemList, name, id } = meal;
    const items = formatItemList(itemList);
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).update({ name, items });

    yield put({
        type: 'EDIT_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
};

function* deleteMeal({ id = '' }) {
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).remove();

    yield put({
        type: 'DELETE_MEAL',
        id,
    });
}

function* setMeal() {
    const meals = [];
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals`).once('value').then((snapshot) => {
        snapshot.forEach((childSnapshot) => {
            const itemList = [];
            childSnapshot.child('items').forEach((item) => {
                itemList.push({
                    itemName: item.key,
                    amount: item.val().amount,
                    unit: item.val().unit,
                });
            });

            meals.push({
                id: childSnapshot.key,
                name: childSnapshot.val().name,
                itemList,
            });
        });
    });

    yield put({ type: 'SET_MEAL', meals });
};

export function* watchAddMeal() {
    yield takeEvery('ON_ADD_MEAL', addMeal);
};

export function* watchEditMeal() {
    yield takeEvery('ON_EDIT_MEAL', editMeal);
};

export function* watchDeleteMeal() {
    yield takeEvery('ON_DELETE_MEAL', deleteMeal);
};

export function* watchSetMeal() {
    yield takeEvery('ON_SET_MEAL', setMeal);
}