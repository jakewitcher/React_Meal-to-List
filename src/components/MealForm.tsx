import React from "react";
import { Item } from "../store/item/types";
import { Meal } from "../store/meal/types";

interface IMealFormProps {
  mealName: string;
  itemName: string;
  amount: string;
  unit: string;
  itemList: [] | Item[];
  mealId: string;
  mealNameChange(e: React.ChangeEvent<HTMLInputElement>): void;
  itemNameChange(e: React.ChangeEvent<HTMLInputElement>): void;
  amountChange(e: React.ChangeEvent<HTMLInputElement>): void;
  unitChange(e: React.ChangeEvent<HTMLSelectElement>): void;
  addItem(): void;
  resetMeal(): void;
  updateMeal(meal: Meal): void;
}

const MealForm: React.FC<IMealFormProps> = props => {
  const {
    mealNameChange,
    itemNameChange,
    amountChange,
    unitChange,
    addItem,
    resetMeal,
    mealName,
    itemName,
    amount,
    unit,
    itemList,
    updateMeal,
    mealId
  } = props;

  const pluralUnitValue = (amount: string, unitType: string) => {
    if (parseFloat(amount) < 2) {
      return `${unitType}`;
    }
    return `${unitType}s`;
  };

  const handleMealChange = (name: string, itemList: Item[], id: string) => {
    updateMeal({
      name,
      itemList,
      id
    });
    resetMeal();
  };

  return (
    <div className="form">
      <div className="input-box input-box--item-name">
        <p>Item</p>
        <input
          className="input"
          type="text"
          value={itemName}
          onChange={itemNameChange}
        />
      </div>

      <div className="input-box input-box--amount">
        <p>Amount</p>
        <input
          className="input"
          type="number"
          value={amount}
          onChange={amountChange}
        />
      </div>

      <div className="input-box input-box--unit">
        <p>Unit</p>
        <select
          className="input input--dropdown"
          name="unit"
          value={unit}
          onChange={unitChange}
        >
          <option value={""}>{"none"}</option>
          <option value={pluralUnitValue(amount, "pound")}>
            {pluralUnitValue(amount, "pound")}
          </option>
          <option value={pluralUnitValue(amount, "ounce")}>
            {pluralUnitValue(amount, "ounce")}
          </option>
          <option value={pluralUnitValue(amount, "gallon")}>
            {pluralUnitValue(amount, "gallon")}
          </option>
          <option value={pluralUnitValue(amount, "quart")}>
            {pluralUnitValue(amount, "quart")}
          </option>
          <option value={pluralUnitValue(amount, "pint")}>
            {pluralUnitValue(amount, "pint")}
          </option>
          <option value={pluralUnitValue(amount, "item")}>
            {pluralUnitValue(amount, "item")}
          </option>
          <option value={pluralUnitValue(amount, "bag")}>
            {pluralUnitValue(amount, "bag")}
          </option>
          <option value={pluralUnitValue(amount, "container")}>
            {pluralUnitValue(amount, "container")}
          </option>
        </select>
      </div>
      <div className="button-box__add">
        <button className="button button__add" onClick={addItem}>
          Add Item
        </button>
      </div>
      <div>
        <div className="input-box__meal-name">
          <p>Meal Name</p>
          <input
            className="input__meal-name"
            type="text"
            value={mealName}
            onChange={mealNameChange}
          />
        </div>
        <div className="button-box__save">
          <button
            className="button button__save"
            onClick={() => handleMealChange(mealName, itemList, mealId)}
          >
            Save Meal
          </button>
        </div>
      </div>
    </div>
  );
};

export default MealForm;
