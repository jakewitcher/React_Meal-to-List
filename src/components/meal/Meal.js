import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { onDeleteMeal } from '../../actions/meal';

export const Meal = (props) => {
    const { name, id, itemList } = props;
    return (
        <div className="list-type">
            <Link className="list-type__link" to={`/meals/meal/${id}`}>
                <p className="list-type__name">{name}</p>
            </Link>
            <p className="list-type__details">number of items: <span className="list-type__details list-type__details--span"> {itemList.length}</span></p>
            <div className="list-type__update">
                <Link className="list-type__link" to={`/meals/edit/${id}`}>
                    <i className="list-type__edit fas fa-edit" />
                </Link>
                <i
                    className="list-type__delete fas fa-trash"
                    onClick={() => props.onDeleteMeal(id)}
                />
            </div>
        </div>
    );
}

const mapDispatchToProps = (dispatch) => {
    return {
        onDeleteMeal: (id) => dispatch(onDeleteMeal(id)),
    }
}

export default connect(undefined, mapDispatchToProps)(Meal);