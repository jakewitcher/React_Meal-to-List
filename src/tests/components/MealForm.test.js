import React from 'react';
import { shallow } from 'enzyme';
import MealForm from '../../components/meal/MealForm';
import { mealItems } from '../fixtures/meal';

let title, 
    mealName, 
    mealNameChange, 
    itemName, 
    itemNameChange, 
    amount, 
    amountChange, 
    unit, 
    unitChange, 
    addItem,
    itemList,
    mealId,
    updateMeal, 
    resetMeal,
    wrapper;

    beforeEach(() => {
        title = 'meal';
        mealName = 'breakfast';
        itemName = 'bacon';
        amount = 5;
        unit = 'pounds';
        itemList = mealItems;
        mealId = 1;
        mealNameChange = jest.fn();
        itemNameChange = jest.fn();
        amountChange = jest.fn();
        unitChange = jest.fn();
        addItem = jest.fn();
        updateMeal = jest.fn();
        resetMeal = jest.fn();
        wrapper = shallow(
            <MealForm 
                title ={title}
                mealName ={mealName}
                itemName={itemName}
                amount={amount}
                unit={unit}
                mealNameChange={mealNameChange}
                itemNameChange={itemNameChange}
                amountChange={amountChange}
                unitChange={unitChange}
                addItem={addItem}
                updateMeal={updateMeal}
                resetMeal={resetMeal}
            />
        );
    });

    test('should render MealForm correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should set meal name', () => {
        const value = { target: { value: mealName }};
        wrapper.find('input').at(0).simulate('change', value);
        expect(mealNameChange).toHaveBeenLastCalledWith(value);
    });

    test('should set item name', () => {
        const value = { target: { value: itemName }};
        wrapper.find('input').at(1).simulate('change', value);
        expect(itemNameChange).toHaveBeenLastCalledWith(value);
    });

    test('should set item amount', () => {
        const value = { target: { value: amount }};
        wrapper.find('input').at(2).simulate('change', value);
        expect(amountChange).toHaveBeenLastCalledWith(value);
    });

    test('should set item unit', () => {
        const value = { target: { value: unit }};
        wrapper.find('select').simulate('change', value);
        expect(unitChange).toHaveBeenLastCalledWith(value);
    });

    test('should add item to meal item list', () => {
        wrapper.find('button').at(0).simulate('click');
        expect(addItem).toHaveBeenCalled();
    });

    test('should add or edit meal and reset state', () => {
        wrapper.find('button').at(1).simulate('click');
        expect(updateMeal).toHaveBeenCalled();
        expect(resetMeal).toHaveBeenCalled();
    });