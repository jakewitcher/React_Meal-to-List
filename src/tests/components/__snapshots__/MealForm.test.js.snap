// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MealForm correctly 1`] = `
<div
  className="form"
>
  <div
    className="input-box input-box--item-name"
  >
    <p>
      Item
    </p>
    <input
      className="input"
      onChange={[MockFunction]}
      type="text"
      value="bacon"
    />
  </div>
  <div
    className="input-box input-box--amount"
  >
    <p>
      Amount
    </p>
    <input
      className="input"
      onChange={[MockFunction]}
      type="number"
      value={5}
    />
  </div>
  <div
    className="input-box input-box--unit"
  >
    <p>
      Unit
    </p>
    <select
      className="input input--dropdown"
      name="unit"
      onChange={[MockFunction]}
      value="pounds"
    >
      <option
        value=""
      >
        none
      </option>
      <option
        value="pounds"
      >
        pounds
      </option>
      <option
        value="ounces"
      >
        ounces
      </option>
      <option
        value="gallons"
      >
        gallons
      </option>
      <option
        value="quarts"
      >
        quarts
      </option>
      <option
        value="pints"
      >
        pints
      </option>
      <option
        value="items"
      >
        items
      </option>
      <option
        value="bags"
      >
        bags
      </option>
      <option
        value="containers"
      >
        containers
      </option>
    </select>
  </div>
  <div
    className="button-box__add"
  >
    <button
      className="button button__add"
      onClick={[MockFunction]}
    >
      Add Item
    </button>
  </div>
  <div>
    <div
      className="input-box__meal-name"
    >
      <p>
        Meal Name
      </p>
      <input
        className="input__meal-name"
        onChange={[MockFunction]}
        type="text"
        value="breakfast"
      />
    </div>
    <div
      className="button-box__save"
    >
      <button
        className="button button__save"
        onClick={[Function]}
      >
        Save Meal
      </button>
    </div>
  </div>
</div>
`;
